cmds/ip/ip.go

```diff
@@ -80,25 +91,71 @@ func dev() netlink.Link {
 	return iface
 }
 
-func showips() {
+func showLinks(w io.Writer, withAddresses bool) {
 	ifaces, err := netlink.LinkList()
 	if err != nil {
 		log.Fatalf("Can't enumerate interfaces? %v", err)
 	}
+
 	for _, v := range ifaces {
-		addrs, err := netlink.AddrList(v, netlink.FAMILY_ALL)
-		if err != nil {
-			log.Printf("Can't enumerate addresses")
+		l := v.Attrs()
+
+		fmt.Fprintf(w, "%d: %s: <%s> mtu %d state %s\n", l.Index, l.Name,
+			strings.Replace(strings.ToUpper(fmt.Sprintf("%s", l.Flags)), "|", ",", -1),
+			l.MTU, strings.ToUpper(l.OperState.String()))
+
+		fmt.Fprintf(w, "    link/%s %s\n", l.EncapType, l.HardwareAddr)
+
+		if withAddresses {
+			showLinkAddresses(w, v)
+		}
+	}
+}
+
+func showLinkAddresses(w io.Writer, link netlink.Link) {
+	addrs, err := netlink.AddrList(link, netlink.FAMILY_ALL)
+	if err != nil {
+		log.Printf("Can't enumerate addresses")
+	}
+
+	for _, addr := range addrs {
+
+		var inet string
+		switch len(addr.IPNet.IP) {
+		case 4:
+			inet = "inet"
+		case 16:
+			inet = "inet6"
+		default:
+			panic("Can't figure out IP protocol version")
```

Comment:
I'm sorry I forgot this but the go documents have a 'don't panic' recommendation. If you can change this to a log.Fatalf before you commit that would help.

Reply:
@rminnich Done.